### DISCUSSION TITLE
AqualinkD and Jandy VSP control

### DISCUSSION THREAD

**Discussion by tziemkus** [aqualinkd/AqualinkD#365] [Q&A] [answered]
First of all I have to say that this is a very impressive project with a lot of dedication and tremendous amount of hours that went into it.

I have been reading the various discussions but can't seem to get my head wrapped around the VSP control comments. @sfeakes mentioned in one of the discussions that it is not possible to control speed settings of a Jandy VSP through AqualinkD if the Aqualink RS panel doesn't support VSP. I have an older panel with revision MM that doesn't seem to have build-in support for VSP.
Currently my VSP is controlled by IQPUMP01. I don't understand why AqualinkD can't issue the required commands over RS485 to control the VSP similar to the commands for turning on e.g. pool light or SPA mode. 
I thought I could disconnect the IQPUMP01 from the VSP, change the DIP switch settings to all off and then connect A+ and A- from the VSP to the Aqualink RS panel RS485.

Is there anybody who can explain why that is not possible?

Thanks in advance for valuable insights....

**Comment by sfeakes** [comment, marked as answer]
I am not sure how IQPUMP01 works, but I will look into it.
To explain this, you need to understand how the Jandy protocol works.
Each device on a RS485 buss has a unique ID, along with only ONE device that can act as a master and all other devices are a slave to that master.   The Jandy Control Panel is FIXED as the master device, and all other devices (keypads,VSP,SWG,Heaters etc)  are slaves.
The master sends out a unique probe messages to EVERY single id within the RS485 range (one by one), if it get's a reply, then the master will start communicating with that device. (ie the slave).   The slaves can ONLY talk to the master, not to each other.
Different keypads (one touch, all button, iAqualinkTouch, RS Serial, PDA etc) use slightly different protocols and are faster at doing different tasks. ie AllButton is the fastest at turning a basic circuit on or off and gathering most statuses, RS Serial faster at turning an advanced circuit on / off, iAqualinkTouch fastest at setting SWG and VSP etc etc.  And different Control Panels support different variations to these keypads, and not all keypads support all functionality.  ie AllButton can't control VSP.

AqualinkD works by acting as multiple slaves (or faking the control panel that it has a few different keypads connected), and depending on what you ask AqualinkD to do, it will used the fastest protocol it has available to perform that task. AqualinkD will also read EVERY message on the RS485 bus and give you instant update as to what's happening.  ie Control Panel sets VSP to something, AqualinkD reads that message and let's it's own UI / MQTT know, before the control panel would have told it's OEM keypad what it did.

So every device on the RS485 bus has to be controlled by the master device (the control panel), and since their is already a master AqualinkD can't take that ID over as that would create a ton of duplicate messages confusing the whole RS485 bus.
Since the VSP is a slave, AqualinkD can't control it directly since there is already a master on the bus.

What I have done in the past, is make AqualinkD the master on a RS485 bus that has no Control Panel.  ie years ago I had a Haywood SWG that the Jandy control panel would not communicate with. So I make a separate RS485 bus that just had a highly modified version of AqualinkD and the SWG on it. (That version of AqualinkD just controlled the SWG), And also had another RS485 bus that had AqualinkD, Control Panel and all other devices on it. (That version of AqualinkD controlled everything else).

So I am assuming that the IQPUMP01 is similar to above, where you simply have the IQPUMP01 and VSP on one physical RS485 bus.  And the control panel / other devices are on a separate physical RS485 bus.   So if that's correct, in theory if you added another USB2RS485 adapter to the PI, modified the AqualinkD code to support two RS485 interfaces, one as master the other as slave, you could probably achieve the same. ie AqualinkD connected to Control Panel as slaves on one physical bus and connected to the VSP as master on a different physical bus.  BUT At present there are some parts of the VSP protocol I have not decoded as there was no need, but to do this you would absolutely have to decode the rest to control the pump fully.

I just found the install manual for IQPUMP01, and it looks like that's exactly how they do it.  It acts as the master device on a RS485 bus that JUST has IQPUMP01 & VSP on it.

**Reply by tziemkus** [reply]
I was hoping for an answer but was not expecting such a detailed and understandable response. Now the brain fog is gone and I think that I now fully understands how this works. 
Thanks a lot for the detailed reply !!!!

**Reply by Bugman1400** [reply]
I also had the same problem. I had the old board with the MMM chip and had to upgrade to the newer board with the daughterboard and T.2 chip. Oddly enough, I went from an RS8 to an RS4 because I found a cheap RS4 board on EBay. Fortunately, all I ever used was 4 outputs so, that worked for me. Otherwise, you're gonna pay at least $800-$1200.

**Reply by sfeakes** [reply]
If you do upgrade, Revision T2 is the one I recommend (if you have a VSP).  It's the best bang for the buck buck off eBay since the price will go by the chip revision, and any revisions newer than T2 have not added any functionality unless you want more than 4 VSP, add pH / ORP sensors or add 20 virtual devices. All the other functionality Jandy has added, (like change RPM / SWG from web app) AqualinkD can do already.

**Reply by tziemkus** [reply]
I finally found a new panel on ebay at a reasonable price ($360). It is a RS8 Combo with version T.1
After changing out the old panel and reconfiguring the aqualinkd.conf/config.js files it was finally working perfectly including the Jandy VSP. However, I experienced some problems with the DIP switch settings inside the pump and setting all DIP switches to off as suggested did not work - so I put them back where the were when using the IQPUMP01 controller. After playing with different pump IDs in the conf file it finally started to work.
Excellent work on the AqualinkD project !!!!!
![image](https://github.com/user-attachments/assets/b076c388-bf16-44fe-a6e5-43f604b80ac7)

Now comes the confusing part. I read a lot about power outages and that it might corrupt the SD card. So I went ahead and enabled OverlayFS with sudo raspi-config and after rebooting the PI started very slowly and then finally after a couple of re-boots it gave up completely leaving me no choice other then starting with a blank SD card from scratch and I finally got it working again.

No idea what went wrong with the OverlayFS option and I am wondering if I did something wrong or should have gone down another route to protect the SD card from being corrupted due to unexpected power loss......

Are there better options than OverlayFS?

**Reply by sfeakes** [reply]
Very glad it's all working.  (On the T.1 vs T.2, I think they are the same in functionally, only that T1 was distributed by a certain distributer)

I personally use a custom RO/root implementation of RaspberryPI os (One I build myself).  OverlayFS should be good, it just has some overhead that I don't like.  There used to be some pre-made RaspberryPI os version with overlayFS, but it seems now the idea is to "enable" it once you have flashed the card.  All I can think is you didn't wait long enough for it to work.

**Reply by tziemkus** [reply]
Believe me, I am also very happy that it works and admire the amount of work you put into it to create such a great application.
I waited 1/2h after I enabled OverlayFS after re-booting and it still didn't work.

Would you mind to share what you did because you clearly know what's best to make it work and avoid a potential SD card corruption.

**Reply by sfeakes** [reply]
To document all the things to make RO root would take more time that I have at the moment, very sorry for that. It’ll also be very dependent upon the OS version using.  But the below blog describes it very well and very similar to what I have done.
https://www.dzombak.com/blog/2024/03/Running-a-Raspberry-Pi-with-a-read-only-root-filesystem.html

I would also say that you need some very good Linux knowledge, and to start with a headless version of Raspberry OS, using one with a UI would probably be a real PITA.

I think Raspberry OS has come a long way in this area over the last few years, so it may not be as much of an issue anymore, Also another easy way is to get everything running, then clone the CF card with something like `dd` that way you can always restore / make a new CF card in a matter of minutes.
